core/src/main/scala/org/http4s/client/jdkhttpclient/WebSocketClient.scala

```diff
@@ -0,0 +1,167 @@
+package org.http4s.client.jdkhttpclient
+
+import cats._
+import cats.data.Chain
+import cats.effect._
+import cats.effect.concurrent.{Deferred, Ref, TryableDeferred}
+import cats.implicits._
+import fs2.{Pipe, Pull, Stream}
+import org.http4s.{Headers, Method, Uri}
+import scodec.bits.ByteVector
+
+/**
+  * A websocket request.
+  *
+  * @param uri The URI.
+  * @param headers The headers to send. Some websocket clients reject protocol-specific headers.
+  * @param method The method of the intial HTTP request. Ignored by some clients.
+  * @param subprotocols The subprotocols, descending from most prefered.
+  */
+case class WSRequest(
+    uri: Uri,
+    headers: Headers = Headers.empty,
+    method: Method = Method.GET,
+    subprotocols: List[String] = List.empty
+)
+
+sealed trait WSFrame
```

Comment:
Do these need to be different from the server model?  It would seem nice to unify.

Reply:
The existing model has some minor issues IMO, so I decided to try sth different in this draft.

 - exposes implementation details (continuation frames should not be part of the public API, you have to parse the close reason manually from a `Close`)
 - `WebSocketFrame` is not sealed
 - creation of a `Close` frame results in an either to prevent invalid close codes - IMHO such validations only result in one more `F.fromEither` and are therefore annoying. I would propose to shift this kind of validation to send time.

An advantage of the server model is that it allows to send text frames with binary data - but I would argue that this is rarely used in practice, and it is supported by neither the JDK ws client nor okhttp (but asynchttpclient allows this).

Reply:
Would the `Close` validation be less annoying if we had a macro validate literal values for the code and/or had adequate constants for the common ones?  Our principle is generally to only allow valid representations at construction, with some concessions to `unsafeFromXXX` methods.  But we do need to keep usability in mind, I agree.

It's difficult when the spec and a subset of implementations support a feature. What might we do if we tried to send such in this or okhttp?  Fail with the web socket equivalent of an UnsupportedOperationException?  If we don't do that, what if we add an even less capable backend later? 🤔

Reply:
Your point concerning different backends is very true - and it also applies to `Close` validation. Quote from `sendClose` from the JDK WS client:

> Status codes 1002, 1003, 1006, 1007, 1009, 1010, 1012, 1013 and 1015 are illegal. Behaviour in respect to other status codes is implementation-specific.

I am unsure about the gains of "irrepresentable illegal states" in this case - IMHO the knowledge "this `send` call might fail for a lot of reasons, but it won't fail because the close code was invalid or the reason was longer than 123 bytes" is not very useful. There is also no parsing etc. involved, which is a difference to e.g. `Uri`.

Also Haskell's `websockets` package does neither validate their WS ADT ([link](https://hackage.haskell.org/package/websockets-0.12.5.3/docs/Network-WebSockets.html#t:Message)) nor provide a validated newtype for close codes ([link](https://hackage.haskell.org/package/websockets-0.12.5.3/docs/Network-WebSockets.html#v:sendCloseCode)) (just as a comparison, I don't want to make an appeal-to-haskthority argument).

Reply:
Some of those JDK WS restrictions make sense as reserved for backend use as opposed to letting an application set them.  I'm comfortable allowing them in the model (they mean something to the receiver!) and having them fail on send.  And once we allow that, you're right, we don't get much out of range validation.  The inconsistency with the rest of the model bothers me, but I find your argument pragmatic.